File: PoketDex/app/src/main/java/pt/rfsfernandes/custom/adapters/PokemonMovesAdapter.java
package pt.rfsfernandes.custom.adapters;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import androidx.core.content.res.ResourcesCompat;
import androidx.recyclerview.widget.RecyclerView;
import pt.rfsfernandes.R;
import pt.rfsfernandes.custom.Constants;
import pt.rfsfernandes.custom.callbacks.MovesItemClick;
import pt.rfsfernandes.custom.utils.UtilsClass;
import pt.rfsfernandes.model.moves.Moves;
import pt.rfsfernandes.model.pokemon_species.FlavourEntries;


public class PokemonMovesAdapter extends RecyclerView.Adapter<PokemonMovesAdapter.MovesViewHolder> {

  private List<Moves> mPokemonMoves = new ArrayList<>();
  private final Context context;
  private final MovesItemClick mMovesItemClick;

  public PokemonMovesAdapter(Context context, MovesItemClick movesItemClick) {
    this.context = context;
    this.mMovesItemClick = movesItemClick;
  }

  @Override
  public MovesViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

    View view = LayoutInflater.from(parent.getContext())
        .inflate(R.layout.moves_row, parent, false);
    return new MovesViewHolder(view);

  }

  @Override
  public void onBindViewHolder(final MovesViewHolder holder, int position) {

    populateItemRows(holder, position);
  }

  /**
   * Populates each row with the content of the current item
   *
   * @param holder   Holder used to populate
   * @param position Position of the item
   */
  private void populateItemRows(MovesViewHolder holder, int position) {

    Moves item = mPokemonMoves.get(position);
    holder.mItem = item;
    holder.textViewMoveTitle.setText(item.getName().replace("-", " "));

    int color = UtilsClass.returnColorId(context, item.getType().getName());

    Drawable drawable = ResourcesCompat.getDrawable(context.getResources(),
        R.drawable.type_background_round,
        null);

    if (drawable != null) {
      drawable.setTint(color);

    }

    holder.textViewMovePower.setBackground(drawable);

    String chosenFlavour = "";
    String country = Locale.getDefault().getLanguage().split("-")[0].toLowerCase();
    for (FlavourEntries flavour :
        item.getFlavourEntriesList()) {
      if (flavour.getLanguage().getName().contains(country)) {
        chosenFlavour = flavour.getFlavourText();

        break;
      }
    }

    if (chosenFlavour.isEmpty()) {
      for (FlavourEntries flavour :
          item.getFlavourEntriesList()) {
        if (flavour.getLanguage().getName().contains("en")) {
          chosenFlavour = flavour.getFlavourText();
          break;
        }
      }
    }

    holder.textViewMoveDescription.setText(chosenFlavour.replace("\n", " "));
    holder.textViewMoveEffect.setText(item.getEffectEntry().getEffect().replace("  ", " ").replace("$effect_chance%", String.valueOf(item.getEffectChance())));
    holder.textViewMoveShortEffect.setText(item.getEffectEntry().getShortEffect().replace("  ",
        " ").replace("$effect_chance%", String.valueOf(item.getEffectChance())));
    holder.textViewMovePower.setText(String.valueOf(item.getPower()));

    holder.textViewMovePower.setOnClickListener(e -> {
      mMovesItemClick.clickedMovesItem("", item.getType().getUrlId(), Constants.MOVES_ITEM.TYPE);
    });

    holder.textViewMoveDescription.setOnClickListener(e -> mMovesItemClick.clickedMovesItem(
        context.getResources().getString(R.string.description),
        holder.textViewMoveDescription.getText().toString(),
        Constants.MOVES_ITEM.SIMPLE));

    holder.textViewMoveEffect.setOnClickListener(e -> mMovesItemClick.clickedMovesItem(
        context.getResources().getString(R.string.effect),
        holder.textViewMoveEffect.getText().toString(),
        Constants.MOVES_ITEM.SIMPLE));

    holder.textViewMoveShortEffect.setOnClickListener(e -> mMovesItemClick.clickedMovesItem(
        context.getResources().getString(R.string.short_effect),
        holder.textViewMoveShortEffect.getText().toString(),
        Constants.MOVES_ITEM.SIMPLE));

  }

  @Override
  public int getItemCount() {
    return mPokemonMoves.size();
  }

  /**
   * Assigns a value to the global List of Moves and notifies the adapter of that change
   *
   * @param moves New list
   */
  public void refreshList(List<Moves> moves) {
    this.mPokemonMoves = moves;
    notifyDataSetChanged();
  }

  public class MovesViewHolder extends RecyclerView.ViewHolder {
    public final View mView;
    public final TextView textViewMoveTitle;
    public final TextView textViewMoveEffect;
    public final TextView textViewMoveShortEffect;
    public final TextView textViewMovePower;
    public final TextView textViewMoveDescription;
    public final View selectedView;
    public Moves mItem;

    public MovesViewHolder(View view) {
      super(view);
      mView = view;
      textViewMoveTitle = view.findViewById(R.id.textViewMoveTitle);

      textViewMoveEffect = view.findViewById(R.id.textViewMoveEffect);
      textViewMoveShortEffect = view.findViewById(R.id.textViewMoveShortEffect);
      textViewMovePower = view.findViewById(R.id.textViewMovePower);
      textViewMoveDescription = view.findViewById(R.id.textViewMoveDescription);

      selectedView = view.findViewById(R.id.viewSelect);
    }

  }
}

File: PocketDex/app/src/main/java/pt/rfsfernandes/pocketdex/custom/adapters/PokemonMovesAdapter.kt
package pt.rfsfernandes.pocketdex.custom.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.RecyclerView
import pt.rfsfernandes.pocketdex.R
import pt.rfsfernandes.pocketdex.custom.Constants
import pt.rfsfernandes.pocketdex.custom.adapters.PokemonMovesAdapter.MovesViewHolder
import pt.rfsfernandes.pocketdex.custom.callbacks.MovesItemClick
import pt.rfsfernandes.pocketdex.custom.utils.UtilsClass
import pt.rfsfernandes.pocketdex.model.moves.Moves
import java.util.*

class PokemonMovesAdapter(
    private val context: Context,
    private val mMovesItemClick: MovesItemClick
) : RecyclerView.Adapter<MovesViewHolder>() {
    private var mPokemonMoves: List<Moves?> = ArrayList()
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovesViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.moves_row, parent, false)
        return MovesViewHolder(view)
    }

    override fun onBindViewHolder(holder: MovesViewHolder, position: Int) {
        populateItemRows(holder, position)
    }

    /**
     * Populates each row with the content of the current item
     *
     * @param holder   Holder used to populate
     * @param position Position of the item
     */
    private fun populateItemRows(holder: MovesViewHolder, position: Int) {
        val item = mPokemonMoves[position]
        holder.mItem = item
        holder.textViewMoveTitle.text = item?.name?.replace("-", " ")
        val color = item?.type?.name?.let { UtilsClass.returnColorId(context, it) }
        val drawable = ResourcesCompat.getDrawable(
            context.resources,
            R.drawable.type_background_round,
            null
        )
        color?.let { drawable?.setTint(it) }
        holder.textViewMovePower.background = drawable
        var chosenFlavour = ""
        val country =
            Locale.getDefault().language.split("-").toTypedArray()[0].lowercase(Locale.getDefault())
        item?.flavourEntriesList?.let {
            for (flavour in it) {
                if (flavour.language.name.contains(country)) {
                    chosenFlavour = flavour.getFlavourText()
                    break
                }
            }
        }

        if (chosenFlavour.isEmpty()) {
            item?.flavourEntriesList?.let {
                for (flavour in it) {
                    if (flavour.language.name.contains("en")) {
                        chosenFlavour = flavour.getFlavourText()
                        break
                    }
                }
            }

        }
        holder.textViewMoveDescription.text = chosenFlavour.replace("\n", " ")
        holder.textViewMoveEffect.text = item?.effectEntry?.effect?.replace("  ", " ")
            ?.replace("\$effect_chance%", item.effectChance.toString())
        holder.textViewMoveShortEffect.text = item?.effectEntry?.shortEffect?.replace(
            "  ",
            " "
        )?.replace("\$effect_chance%", item.effectChance.toString())
        holder.textViewMovePower.text = item?.power.toString()
        holder.textViewMovePower.setOnClickListener { e: View? ->
            mMovesItemClick.clickedMovesItem(
                "",
                item?.type?.urlId,
                Constants.MOVES_ITEM.TYPE
            )
        }
        holder.textViewMoveDescription.setOnClickListener { e: View? ->
            mMovesItemClick.clickedMovesItem(
                context.resources.getString(R.string.description),
                holder.textViewMoveDescription.text.toString(),
                Constants.MOVES_ITEM.SIMPLE
            )
        }
        holder.textViewMoveEffect.setOnClickListener { e: View? ->
            mMovesItemClick.clickedMovesItem(
                context.resources.getString(R.string.effect),
                holder.textViewMoveEffect.text.toString(),
                Constants.MOVES_ITEM.SIMPLE
            )
        }
        holder.textViewMoveShortEffect.setOnClickListener { e: View? ->
            mMovesItemClick.clickedMovesItem(
                context.resources.getString(R.string.short_effect),
                holder.textViewMoveShortEffect.text.toString(),
                Constants.MOVES_ITEM.SIMPLE
            )
        }
    }

    override fun getItemCount(): Int {
        return mPokemonMoves.size
    }

    /**
     * Assigns a value to the global List of Moves and notifies the adapter of that change
     *
     * @param moves New list
     */
    fun refreshList(moves: List<Moves?>) {
        mPokemonMoves = moves
        notifyDataSetChanged()
    }

    inner class MovesViewHolder(val mView: View) : RecyclerView.ViewHolder(mView) {
        val textViewMoveTitle: TextView = mView.findViewById(R.id.textViewMoveTitle)
        val textViewMoveEffect: TextView = mView.findViewById(R.id.textViewMoveEffect)
        val textViewMoveShortEffect: TextView = mView.findViewById(R.id.textViewMoveShortEffect)
        val textViewMovePower: TextView = mView.findViewById(R.id.textViewMovePower)
        val textViewMoveDescription: TextView = mView.findViewById(R.id.textViewMoveDescription)
        var mItem: Moves? = null
    }
}